ir/base/reachable.go

```diff
@@ -70,166 +84,128 @@ func (r *Reachable) UpdateWith(ctx ir.UpdateContext, with ir.Node) error {
-	d, dok := p.Positional[0].(xr.Dict)
-	s, sok := p.Positional[0].(xr.List)
-	if !dok && !sok {
-		return nil, fmt.Errorf("expecting dict or list")
-	}
-	// Get tag from predicate
+	// Get tag and positional arguments.
 	tag := p.Tag
-	if dok {
-		return reachableDictAssemble(ctx, tag, d, metadata...)
-	}
-	return reachableListAssemble(ctx, tag, s, metadata...)
-
-}
-
-func reachableDictAssemble(ctx ir.AssemblerContext, tag string, d xr.Dict, metadata ...ir.Metadata) (ir.Node, error) {
-
-	isConn := false
-	if tag != "connected" && tag != "dialable" {
-		return nil, fmt.Errorf("expecting tag 'connected' or 'dialable'")
-	}
-	// If the node is of type connected List flag
-	if tag == "connected" {
-		isConn = true
+	addr := getNamed(p, xr.String{"address"})
+	// Check tag
+	if tag != "connectivity" && tag != "dialable" {
+		return nil, fmt.Errorf("not a reachable smart tag")
 	}
 
-	u := xr.Dict{}
-	r := xr.Dict{}
-	for _, p := range d.Pairs {
-		info := isValidMultiAddrNode(p.Value)
-		// If not a multiaddr add to user List and continue
-		if info == nil {
-			// Add non-multiaddr to user-dict
-			u.Pairs = append(u.Pairs, p)
-			continue
-		}
-		// According to if connected or dialable
-		if isConn {
-			// If connected add pair with multiaddr to reachable
-			if conn := checkIfConnected(ctx.Host, *info); conn {
-				r.Pairs = append(r.Pairs, p)
-			}
-		} else {
-			// If dialable add pair with multiaddr to reachable
-			if dialable := checkIfDialable(ctx.Host, *info); dialable {
-				r.Pairs = append(r.Pairs, p)
-			}
-		}
-	}
-	// Assemble reachable and user dicts.
-	asm := ir.DictAssembler{}
-	uasm, err := asm.Assemble(ctx, u, metadata...)
+	// Check multiaddress
+	maddr, err := parse.ParseMultiaddr(&parse.ParseCtx{}, addr)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't assemble user dict: %s", err)
+		return nil, fmt.Errorf("no valid multiaddr provided")
 	}
-	rasm, err := asm.Assemble(ctx, r, metadata...)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't assemble reachable dict: %s", err)
+
+	// Assemble metadata
+	m := meta.New()
+	if err := m.Apply(metadata...); err != nil {
+		return nil, err
 	}
+
 	return &Reachable{
-		Reachable: rasm,
-		User:      uasm,
-		isConn:    isConn,
+		addr:        maddr,
+		verifyConn:  tag == "connectivity",
+		verifyDial:  tag == "dialable",
+		metadataCtx: m,
 	}, nil
 }
 
-func reachableListAssemble(ctx ir.AssemblerContext, tag string, d xr.List, metadata ...ir.Metadata) (ir.Node, error) {
-	isConn := false
-	if tag != "connected" && tag != "dialable" {
-		return nil, fmt.Errorf("expecting tag 'connected' or 'dialable'")
-	}
-	// If the node is of type connected List flag
-	if tag == "connected" {
-		isConn = true
+// TriggerReachable triggers the execution of Reachable verifications
+// over a dict and adds the appropiate flag to Nodes that don't pass the verification.
+func TriggerReachable(d *ir.Dict, h host.Host) {
+	// For each pair.
+	for k := len(d.Pairs) - 1; k >= 0; k-- {
```

Comment:
Why is the iteration going backwards? Couldn't you just use:
for _, p := d.Pairs {
  ...
}

Reply:
Good point! In the previous implementation, non-reachable nodes were being removed in each iteration (and thus the backward loop). Now we don't need it but I forgot to change it. Now is fixed. Thanks!